Cesium3DTilesSelection/src/LayerJsonTerrainLoader.cpp

```diff
@@ -0,0 +1,725 @@
+              return std::to_string(tileID.x);
+            }
+            if (placeholder == "y") {
+              return std::to_string(tileID.y);
+            }
+            if (placeholder == "version") {
+              return layer.version;
+            }
+
+            return placeholder;
+          }));
+}
+
+Future<TileLoadResult> requestTileContent(
+    const std::shared_ptr<spdlog::logger>& pLogger,
+    const AsyncSystem& asyncSystem,
+    const std::shared_ptr<IAssetAccessor>& pAssetAccessor,
+    const QuadtreeTileID& tileID,
+    const BoundingVolume& boundingVolume,
+    LayerJsonTerrainLoader::Layer& layer,
+    const std::vector<IAssetAccessor::THeader>& requestHeaders,
+    bool enableWaterMask) {
+  std::string url = resolveTileUrl(tileID, layer);
+
+  Future<QuantizedMeshLoadResult> futureQuantizedMesh =
+      pAssetAccessor->get(asyncSystem, url, requestHeaders)
+          .thenInWorkerThread(
+              [asyncSystem, pLogger, tileID, boundingVolume, enableWaterMask](
+                  std::shared_ptr<IAssetRequest>&& pRequest) {
+                const IAssetResponse* pResponse = pRequest->response();
+                if (!pResponse) {
+                  QuantizedMeshLoadResult result;
+                  result.errors.emplace_error(fmt::format(
+                      "Did not receive a valid response for tile content {}",
+                      pRequest->url()));
+                  return result;
+                }
+
+                if (pResponse->statusCode() != 0 &&
+                    (pResponse->statusCode() < 200 ||
+                     pResponse->statusCode() >= 300)) {
+                  QuantizedMeshLoadResult result;
+                  result.errors.emplace_error(fmt::format(
+                      "Received status code {} for tile content {}",
+                      pResponse->statusCode(),
+                      pRequest->url()));
+                }
+
+                return QuantizedMeshLoader::load(
+                    tileID,
+                    boundingVolume,
+                    pRequest->url(),
+                    pResponse->data(),
+                    enableWaterMask);
+              });
+
+  // If this tile has availability data, we need to add it to the layer in the
+  // main thread.
+  if (layer.availabilityLevels > 0 &&
+      int32_t(tileID.level) % layer.availabilityLevels == 0) {
+    futureQuantizedMesh =
+        std::move(futureQuantizedMesh)
+            .thenInMainThread(
+                [&layer](QuantizedMeshLoadResult&& loadResult) mutable {
+                  for (auto& range : loadResult.availableTileRectangles) {
+                    layer.availability.addAvailableTileRange(range);
+                  }
+                  return std::move(loadResult);
+                });
+  }
+
+  return std::move(futureQuantizedMesh)
+      .thenImmediately([](QuantizedMeshLoadResult&& loadResult) {
+        // TODO: why does TileLoadRequest need an IAssetRequest?
```

Comment:
I think I see why now: it's needed to determine that a load failure was the result of a 401, so we know to refresh the ion token (if this is an ion request). Getting the request into this lambda is super awkward, though.